src/rk4.rs

```diff
@@ -62,47 +89,74 @@ where
 
         let num_steps = ((self.x_end - self.x) / self.step_size).ceil() as usize;
         for _ in 0..num_steps {
-            let (x_new, y_new) = self.step();
-
-            self.x_out.push(x_new);
-            self.y_out.push(y_new.clone());
+            let (x_new, y_new, abort) = self.step();
 
             self.x = x_new;
-            self.y = y_new;
+            self.y
+                .iter_mut()
+                .zip(y_new.iter())
+                .for_each(|(y_self, y_new_elem)| {
+                    *y_self = *y_new_elem;
+                });
+
+            self.x_out.push(x_new);
+            self.y_out.push(y_new);
 
             self.stats.num_eval += 4;
             self.stats.accepted_steps += 1;
+
+            if abort {
+                break;
+            }
         }
         Ok(self.stats)
     }
 
     /// Performs one step of the Runge-Kutta 4 method.
-    fn step(&self) -> (f64, OVector<T, D>) {
-        let (rows, cols) = self.y.shape_generic();
-        let mut k = vec![OVector::zeros_generic(rows, cols); 12];
-
-        self.f.system(self.x, &self.y, &mut k[0]);
-        self.f.system(
-            self.x + self.half_step,
-            &(self.y.clone() + k[0].clone() * self.half_step),
-            &mut k[1],
-        );
-        self.f.system(
-            self.x + self.half_step,
-            &(self.y.clone() + k[1].clone() * self.half_step),
-            &mut k[2],
-        );
-        self.f.system(
-            self.x + self.step_size,
-            &(self.y.clone() + k[2].clone() * self.step_size),
-            &mut k[3],
-        );
+    fn step(&mut self) -> (f64, OVector<T, D>, bool) {
+        self.f.system(self.x, &self.y, &mut self.k[0]);
+
+        self.populate_buffer(0, self.half_step);
+        self.f
+            .system(self.x + self.half_step, &self.buffer, &mut self.k[1]);
+
+        self.populate_buffer(1, self.half_step);
+        self.f
+            .system(self.x + self.half_step, &self.buffer, &mut self.k[2]);
+
+        self.populate_buffer(2, self.step_size);
+        self.f
+            .system(self.x + self.step_size, &self.buffer, &mut self.k[3]);
 
         let x_new = self.x + self.step_size;
-        let y_new = &self.y
-            + (k[0].clone() + k[1].clone() * 2.0 + k[2].clone() * 2.0 + k[3].clone())
-                * (self.step_size / 6.0);
-        (x_new, y_new)
+
+        let mut y_new = self.y.clone();
+
+        for (idx, y_elem) in y_new.iter_mut().enumerate() {
+            *y_elem = *y_elem
+                + (self.k[0][idx]
+                    + self.k[1][idx]
+                    + self.k[1][idx]
```

Comment:
Why not `+ 2. * self.k[1][idx]`? (same for k[2])

Reply:
To be honest: I have no idea why I did it that way :-D If you prefer, I can change it to your suggestion.

Reply:
No worries, sorry I took so long to review it. Yes it would be great if you can update this and rebase on main so that I can merge your next revision.